### ISSUE TITLE
systemd socket and service conflict?

### ISSUE THREAD

**Issue by volkriss** [ipfs/go-ipfs#8905]
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [go-ipfs version](https://dist.ipfs.io/#go-ipfs) or have an issue updating.

### Installation method

built from source

### Version

```Text
go-ipfs version: 0.4.17-
Repo version: 7
System version: amd64/linux
Golang version: go1.10.3
```

### Description

When I installed the systemd .service and .socket files from [the systemd directory](https://github.com/ipfs/go-ipfs/tree/master/misc/systemd) as system files, the sockets prevented the service from starting since they bound up the listening ports. journalctl showed systemd trying over and over to start go-ipfs, always failing with that error.

This meant go-ipfs did not restart on reboot.

I don't understand systemd well enough to fully troubleshoot this.

Maybe there's an error in the provided files? Or maybe I shouldn't have installed the socket files if I'm installing the service as a system service instead of a user service? In that case, it would be nice if the documentation in the parent directory had a note warning not to install the files.

It sounds to me like the .socket files were expected to work, though, since #6675 talks about socket activation.

Either way, whether it's user error or something needing tweaking in the systemd files, I wanted to report this issue getting the daemon going on startup.

**Comment by Stebalien** [comment]
Socket activation was added in 0.5.1 which was released in 2020. You're using 4.17, which was released in 2018. Please update.

**Comment by volkriss** [comment]
I have no idea where the 0.4.17 came from, but the web interface says I'm running 0.12.2, consistent with my recent download of the package.

go-ipfs version: 0.12.2
Repo version: 12
System version: amd64/linux
Golang version: go1.17.8

**Comment by Stebalien** [comment]
Maybe you have multiple versions installed?

What's the exact error message?

**Comment by Stebalien** [comment]
And have you changed the config?

**Comment by volkriss** [comment]
This report should probably be paused because:

Yeah, this version inconsistency is strange, so when I have a second I need to figure it out.
I'm building from source from tagged releases here on github, so I'm thinking something in my build went wrong.

ipfs --api /ip4/127.0.0.1/tcp/5001 version --all
says: go-ipfs version: 0.4.17-

wget "http://127.0.0.1:8000/api/v0/version?enc=text&all=true"
says: ipfs version 0.12.2

**Comment by Stebalien** [comment]
Are you using IPFS desktop? Brave?

**Comment by Ericson2314** [comment]
On NixOS I have at times seen some oddities, but I have other other socket-activated programs too so I am inclined to think it might just be systemd's fault.

**Comment by volkriss** [comment]
Like I said, I'm going to track down this version issue when I have a second, maybe tonight.
I have a theory to follow up on, and I'll report back when I look into it.

I do suspect @Stebalien is right, that there's no bug here, just something needing updating.

But just for a quick test, I see one copy of go-ipfs, confirmed by showing the same PID listening on both sockets:
~$ sudo netstat --inet -l -p | grep ipfs
tcp        0      0 0.0.0.0:5001            0.0.0.0:*               LISTEN      65635/ipfs
tcp        0      0 0.0.0.0:8000            0.0.0.0:*               LISTEN      65635/ipfs

**Comment by Ericson2314** [comment]
@Stebalien Sorry I dropped a key part of the sentence where the same issues with socket activation and manually restarting stuff I see with all socket activated services alike. Agree there doesn't seem to be any go-ipfs issue.

**Comment by volkriss** [comment]
I worked out the version issue. Lesson learned: ipfs --api xxx version reports the version of _that_ ipfs and not the daemon it's pointed to through the api.

Now that I'm sure I'm running 0.12.2, I'm reproducing the issue.

On startup systemd activates the socket files and then shows ipfs repeatedly trying but failing to launch with this error:
ipfs[1000]: Error: serveHTTPApi: manet.Listen(/ip4/0.0.0.0/tcp/5001) failed: listen tcp4 0.0.0.0:5001: bind: address already in use
systemd[1]: ipfs.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: ipfs.service: Failed with result 'exit-code'.
systemd[1]: Failed to start InterPlanetary File System (IPFS) daemon.

Once I stop the socket files, though, ipfs starts successfully on systemd's next try to start it:

~$: sudo systemctl stop ipfs-api.socket ipfs-gateway.socket

systemd[1]: ipfs-api.socket: Succeeded.
systemd[1]: Closed IPFS API Socket.
systemd[1]: ipfs-gateway.socket: Succeeded.
systemd[1]: Closed IPFS Gateway Socket.

ipfs[2919]: API server listening on /ip4/0.0.0.0/tcp/5001
ipfs[2919]: WebUI: http://0.0.0.0:5001/webui
ipfs[2919]: Gateway (readonly) server listening on /ip4/0.0.0.0/tcp/8000
ipfs[2919]: Daemon is ready
systemd[1]: Started InterPlanetary File System (IPFS) daemon.

The only changes I can think of were moving the gateway listen port to 8000 instead of 8080 in both the socket file and ipfs config, setting listen ports to allow access from the outside, and --enable-pubsub-experiment in ipfs.service ExecStart.

**Comment by Stebalien** [comment]
Oh, I see what's happening. You've changed the config to listen on `0.0.0.0`, but the socket is already listening on `127.0.0.1`. Right?

**Comment by Stebalien** [comment]
> I would suggest adding a note about keeping *.socket consistent with ipfs config to misc/README.md, and I'm happy to propose an edit to that document if you think it's appropriate.

Yeah, this should definitely be documented. We could _probably_ have a better error as well, but it's a bit tricky.

Basically:

1. The config & systemd can specify the same addresses.
2. The config & systemd can specify different non-conflicting addresses.
3. They cannot specify _different_ addresses that conflict.

For example:

- You could listen on 127.0.0.1:5001 with systemd.
- The config can specify 1.2.3.4:5001 (different IP), 127.0.0.1:5001 (identical), 127.0.0.1:1234 (different port), 0.0.0.0:1234 (all address, but different port).
